File: VIDEO/app/blueprints/pending_enroll.py
"""待入库工作台路由：未匹配人脸/车牌目标的批量管理与确认入库。

人脸线与车牌线是两条独立业务线，路由同构：
  /video/face/pending-enroll/...   人脸（特征向量入库）
  /video/plate/pending-enroll/...  车牌（车牌号入库）
"""

from flask import Blueprint, Response, current_app, jsonify, request

from app.services import pending_enroll_service


def create_pending_enroll_blueprint(kind: str) -> Blueprint:
    bp = Blueprint(f"pending_enroll_{kind}", __name__)

    def _error(msg: str, code: int = 400):
        return jsonify({"code": code, "msg": msg}), code

    @bp.route("/records", methods=["GET"])
    def list_pending_records():
        try:
            page = int(request.args.get("page", 1))
            page_size = int(request.args.get("pageSize", request.args.get("page_size", 24)))
            data = pending_enroll_service.list_records(
                kind,
                status=request.args.get("status", "pending") or "pending",
                page=page,
                page_size=page_size,
                search=request.args.get("search", "").strip() or None,
                device_id=request.args.get("device_id", "").strip() or None,
                task_id=int(request.args["task_id"]) if request.args.get("task_id") else None,
            )
            return jsonify({"code": 0, "msg": "success", **data})
        except ValueError as exc:
            return _error(str(exc))
        except Exception as exc:
            current_app.logger.error(f"查询待入库记录失败: {exc}", exc_info=True)
            return _error(f"查询失败: {exc}", 500)

    @bp.route("/stats", methods=["GET"])
    def pending_stats():
        try:
            return jsonify({"code": 0, "msg": "success", "data": pending_enroll_service.get_stats(kind)})
        except Exception as exc:
            current_app.logger.error(f"查询待入库统计失败: {exc}", exc_info=True)
            return _error(f"查询失败: {exc}", 500)

    @bp.route("/records/<int:record_id>", methods=["GET"])
    def get_pending_record(record_id: int):
        try:
            data = pending_enroll_service.get_record_detail(kind, record_id)
            return jsonify({"code": 0, "msg": "success", "data": data})
        except ValueError as exc:
            return _error(str(exc))
        except Exception as exc:
            current_app.logger.error(f"查询待入库记录详情失败: {exc}", exc_info=True)
            return _error(f"查询失败: {exc}", 500)

    @bp.route("/records/batch-discard", methods=["POST"])
    def batch_discard_records():
        try:
            data = request.get_json(silent=True) or {}
            ids = data.get("ids") or data.get("record_ids") or []
            if not isinstance(ids, list) or not ids:
                return _error("ids 不能为空")
            changed = pending_enroll_service.discard_records(kind, [int(x) for x in ids])
            return jsonify({"code": 0, "msg": f"已忽略 {changed} 条", "data": {"changed": changed}})
        except ValueError as exc:
            return _error(str(exc))
        except Exception as exc:
            current_app.logger.error(f"批量忽略失败: {exc}", exc_info=True)
            return _error(f"操作失败: {exc}", 500)

    @bp.route("/records/batch-restore", methods=["POST"])
    def batch_restore_records():
        try:
            data = request.get_json(silent=True) or {}
            ids = data.get("ids") or data.get("record_ids") or []
            if not isinstance(ids, list) or not ids:
                return _error("ids 不能为空")
            changed = pending_enroll_service.restore_records(kind, [int(x) for x in ids])
            return jsonify({"code": 0, "msg": f"已恢复 {changed} 条", "data": {"changed": changed}})
        except ValueError as exc:
            return _error(str(exc))
        except Exception as exc:
            current_app.logger.error(f"批量恢复失败: {exc}", exc_info=True)
            return _error(f"操作失败: {exc}", 500)

    @bp.route("/records/batch-delete", methods=["POST"])
    def batch_delete_records():
        try:
            data = request.get_json(silent=True) or {}
            ids = data.get("ids") or data.get("record_ids") or []
            if not isinstance(ids, list) or not ids:
                return _error("ids 不能为空")
            deleted = pending_enroll_service.delete_records(kind, [int(x) for x in ids])
            return jsonify({"code": 0, "msg": f"已删除 {deleted} 条", "data": {"deleted": deleted}})
        except ValueError as exc:
            return _error(str(exc))
        except Exception as exc:
            current_app.logger.error(f"批量删除失败: {exc}", exc_info=True)
            return _error(f"操作失败: {exc}", 500)

    @bp.route("/records/<int:record_id>/extract-preview", methods=["POST"])
    def extract_preview(record_id: int):
        """按（修正后的）标注框返回提取区域 JPEG 预览"""
        try:
            data = request.get_json(silent=True) or {}
            crop_bytes, info = pending_enroll_service.extract_preview_bytes(
                kind, record_id, bbox=data.get("bbox"),
            )
            return Response(crop_bytes, mimetype="image/jpeg", headers={
                "X-Extract-Source": info.get("source", ""),
            })
        except ValueError as exc:
            return _error(str(exc))
        except Exception as exc:
            current_app.logger.error(f"提取预览失败: {exc}", exc_info=True)
            return _error(f"提取失败: {exc}", 500)

    @bp.route("/records/<int:record_id>/enroll", methods=["POST"])
    def enroll_record(record_id: int):
        try:
            payload = request.get_json(silent=True) or {}
            result = pending_enroll_service.enroll_record(kind, record_id, payload)
            msg = (
                f"已录入 {result['entry'].get('person_name', '')}"
                if kind == "face"
                else f"已录入车牌 {result['entry'].get('plate_no', '')}"
            )
            return jsonify({"code": 0, "msg": msg, "data": result})
        except ValueError as exc:
            return _error(str(exc))
        except Exception as exc:
            current_app.logger.error(f"待入库记录入库失败: {exc}", exc_info=True)
            return _error(f"入库失败: {exc}", 500)

    @bp.route("/records/batch-enroll", methods=["POST"])
    def batch_enroll_records():
        try:
            data = request.get_json(silent=True) or {}
            items = data.get("items") or []
            if not isinstance(items, list) or not items:
                return _error("items 不能为空")
            result = pending_enroll_service.batch_enroll_records(kind, items)
            msg = f"成功入库 {result['success_count']} 条"
            if result["failed_count"]:
                msg += f"，{result['failed_count']} 条失败"
            return jsonify({"code": 0, "msg": msg, "data": result})
        except ValueError as exc:
            return _error(str(exc))
        except Exception as exc:
            current_app.logger.error(f"批量入库失败: {exc}", exc_info=True)
            return _error(f"批量入库失败: {exc}", 500)

    return bp


pending_enroll_face_bp = create_pending_enroll_blueprint(pending_enroll_service.KIND_FACE)
pending_enroll_plate_bp = create_pending_enroll_blueprint(pending_enroll_service.KIND_PLATE)

File: WEB/src/api/device/pending_enroll.ts
/**
 * 待入库工作台接口（未匹配人脸/车牌目标的暂存、标注修正与确认入库）
 *
 * 人脸线与车牌线为两条独立业务线，接口同构：
 *  - face  → /video/face/pending-enroll/...（特征向量入库）
 *  - plate → /video/plate/pending-enroll/...（车牌号入库）
 */
import { defHttp } from '@/utils/http/axios';
import { parseFaceApiError } from '@/api/device/face_library';

export type PendingEnrollKind = 'face' | 'plate';

export type PendingEnrollStatus = 'pending' | 'enrolled' | 'discarded';

const PREFIX_MAP: Record<PendingEnrollKind, string> = {
  face: '/video/face/pending-enroll',
  plate: '/video/plate/pending-enroll',
};

const SUCCESS_CODES = new Set([0, 200]);

/** AI 检测框（整帧像素坐标） */
export interface PendingEnrollBbox {
  x: number;
  y: number;
  w: number;
  h: number;
}

export interface PendingEnrollStats {
  pending: number;
  enrolled: number;
  discarded: number;
}

export interface PendingEnrollRecord {
  id: number;
  task_id?: number;
  task_name?: string;
  device_id: string;
  device_name?: string;
  /** 人脸线：OCR/匹配信息 */
  plate_no?: string;
  plate_color?: string;
  detect_conf?: number;
  similarity?: number;
  matched?: boolean;
  matched_person_name?: string;
  enroll_status: PendingEnrollStatus;
  /** AI 标注框 [x1, y1, x2, y2]（整帧像素坐标） */
  bbox?: [number, number, number, number] | null;
  frame_image_path?: string | null;
  frame_image_url?: string | null;
  /** 整帧是否仍可用（过期后只能基于裁剪图入库，无法修正标注框） */
  frame_available: boolean;
  frame_width?: number;
  frame_height?: number;
  crop_image_path?: string | null;
  crop_image_url?: string | null;
  enroll_target_library_id?: number | null;
  enroll_person_id?: number | null;
  enroll_entry_id?: number | null;
  enroll_time?: string | null;
  created_at?: string;
}

export interface PendingEnrollListResult {
  code: number;
  msg: string;
  list: PendingEnrollRecord[];
  total: number;
  page: number;
  page_size: number;
  stats: PendingEnrollStats;
}

export interface PendingEnrollListParams {
  status?: PendingEnrollStatus | 'all';
  page?: number;
  pageSize?: number;
  search?: string;
  device_id?: string;
  task_id?: number;
}

function prefixOf(kind: PendingEnrollKind) {
  return PREFIX_MAP[kind];
}

async function mutationApi<T = unknown>(
  kind: PendingEnrollKind,
  url: string,
  data?: unknown,
): Promise<{ code: number; msg: string; data?: T }> {
  defHttp.setHeader({ 'X-Authorization': 'Bearer ' + localStorage.getItem('jwt_token') });
  try {
    const res = await defHttp.post(
      {
        url: `${prefixOf(kind)}${url}`,
        data,
        headers: {
          // @ts-ignore
          ignoreCancelToken: true,
        },
      },
      { isTransformResponse: false, errorMessageMode: 'none' },
    );
    const body = ((res as { data?: { code: number; msg: string; data?: T } })?.data ?? res) as {
      code: number;
      msg: string;
      data?: T;
    };
    if (!SUCCESS_CODES.has(body?.code)) {
      throw new Error(body?.msg || '请求失败');
    }
    return body;
  } catch (error) {
    throw new Error(parseFaceApiError(error, '请求失败'));
  }
}

export const listPendingRecords = async (
  kind: PendingEnrollKind,
  params?: PendingEnrollListParams,
): Promise<PendingEnrollListResult> => {
  defHttp.setHeader({ 'X-Authorization': 'Bearer ' + localStorage.getItem('jwt_token') });
  const res = await defHttp.get<PendingEnrollListResult>(
    {
      url: `${prefixOf(kind)}/records`,
      params,
      headers: {
        // @ts-ignore
        ignoreCancelToken: true,
      },
    },
    { isTransformResponse: false, errorMessageMode: 'none' },
  );
  // isTransformResponse:false 时 defHttp 返回完整 AxiosResponse，业务体在其 data 字段
  return (((res as { data?: PendingEnrollListResult })?.data ?? res) as PendingEnrollListResult);
};

export const getPendingStats = async (
  kind: PendingEnrollKind,
): Promise<{ code: number; msg: string; data: PendingEnrollStats }> => {
  defHttp.setHeader({ 'X-Authorization': 'Bearer ' + localStorage.getItem('jwt_token') });
  const res = await defHttp.get<{ code: number; msg: string; data: PendingEnrollStats }>(
    {
      url: `${prefixOf(kind)}/stats`,
      headers: {
        // @ts-ignore
        ignoreCancelToken: true,
      },
    },
    { isTransformResponse: false, errorMessageMode: 'none' },
  );
  return (((res as { data?: { code: number; msg: string; data: PendingEnrollStats } })?.data ??
    res) as { code: number; msg: string; data: PendingEnrollStats });
};

export const getPendingRecordDetail = async (
  kind: PendingEnrollKind,
  recordId: number,
): Promise<{ code: number; msg: string; data: PendingEnrollRecord }> => {
  defHttp.setHeader({ 'X-Authorization': 'Bearer ' + localStorage.getItem('jwt_token') });
  const res = await defHttp.get<{ code: number; msg: string; data: PendingEnrollRecord }>(
    {
      url: `${prefixOf(kind)}/records/${recordId}`,
      headers: {
        // @ts-ignore
        ignoreCancelToken: true,
      },
    },
    { isTransformResponse: false, errorMessageMode: 'none' },
  );
  return (((res as { data?: { code: number; msg: string; data: PendingEnrollRecord } })?.data ??
    res) as { code: number; msg: string; data: PendingEnrollRecord });
};

export const batchDiscardPendingRecords = (kind: PendingEnrollKind, ids: number[]) =>
  mutationApi<{ changed: number }>(kind, '/records/batch-discard', { ids });

export const batchRestorePendingRecords = (kind: PendingEnrollKind, ids: number[]) =>
  mutationApi<{ changed: number }>(kind, '/records/batch-restore', { ids });

export const batchDeletePendingRecords = (kind: PendingEnrollKind, ids: number[]) =>
  mutationApi<{ deleted: number }>(kind, '/records/batch-delete', { ids });

/**
 * 按修正后的标注框请求提取区域 JPEG 预览，返回对象 URL（调用方负责 revoke）。
 */
export const fetchExtractPreviewUrl = async (
  kind: PendingEnrollKind,
  recordId: number,
  bbox?: [number, number, number, number] | null,
): Promise<string> => {
  defHttp.setHeader({ 'X-Authorization': 'Bearer ' + localStorage.getItem('jwt_token') });
  const [x1, y1, x2, y2] = bbox ?? [];
  const blob = (await defHttp.post(
    {
      url: `${prefixOf(kind)}/records/${recordId}/extract-preview`,
      data: bbox?.length === 4 ? { bbox: [x1, y1, x2, y2] } : {},
      responseType: 'blob',
      headers: {
        // @ts-ignore
        ignoreCancelToken: true,
      },
    },
    { isTransformResponse: false, errorMessageMode: 'none' },
  )) as unknown as Blob;
  if (!(blob instanceof Blob)) {
    throw new Error('提取预览失败');
  }
  return URL.createObjectURL(blob);
};

/** 单条确认入库入参 */
export interface PendingEnrollPayload {
  record_id: number;
  /** 目标库 ID */
  library_id: number;
  /** 标注框（整帧像素坐标），不传则使用 AI 原始框 */
  bbox?: [number, number, number, number] | null;
  // 人脸线字段
  person_id?: number;
  person_name?: string;
  person_code?: string;
  // 车牌线字段
  plate_no?: string;
  plate_color?: string;
  owner_name?: string;
  owner_phone?: string;
  remark?: string;
}

export const enrollPendingRecord = (kind: PendingEnrollKind, payload: PendingEnrollPayload) => {
  const { record_id, ...rest } = payload;
  return mutationApi(kind, `/records/${record_id}/enroll`, rest);
};

export const batchEnrollPendingRecords = (
  kind: PendingEnrollKind,
  items: PendingEnrollPayload[],
) =>
  mutationApi<{
    success_count: number;
    failed_count: number;
    success: Array<{ record_id: number; entry: Record<string, unknown> }>;
    failed: Array<{ record_id?: number; msg: string }>;
  }>(kind, '/records/batch-enroll', { items });

/** AI 框 [x1,y1,x2,y2] → 编辑器矩形 {x,y,w,h} */
export function bboxToRect(
  bbox?: [number, number, number, number] | null,
): PendingEnrollBbox | null {
  if (!bbox || bbox.length < 4) return null;
  const [x1, y1, x2, y2] = bbox;
  const rect = { x: Math.min(x1, x2), y: Math.min(y1, y2), w: Math.abs(x2 - x1), h: Math.abs(y2 - y1) };
  if (rect.w <= 0 || rect.h <= 0) return null;
  return rect;
}

/** 编辑器矩形 → [x1,y1,x2,y2] */
export function rectToBbox(rect: PendingEnrollBbox): [number, number, number, number] {
  return [
    Math.round(rect.x),
    Math.round(rect.y),
    Math.round(rect.x + rect.w),
    Math.round(rect.y + rect.h),
  ];
}
